HBASE-29228 Add support to prevent running multiple active clusters

kabhishek4 opened the pull request:
Using HBase Active and Read Replica Clusters, there could be potential problem scenarios where multiple clusters started in the active mode. This change set prevents starting multiple clusters in active mode.

anmolnar reviewed (changes requested):
Looking at the code I think this approach is not going to work. There're two ways to determine the cluster id for each cluster: read it from the FS or if file is missing, generate a new one. In this case, since the storage location is the same, the filesystem is shared between the active cluster and read only clusters, all clusters will have the same cluster id. There's no point to do the comparison at startup, because cluster IDs will always be the same.

Checking the presence of the `active_cluster_id` file still makes sense, but if the active cluster is unable to remove it before it shuts down, it won't be able to start again.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 42s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  |  Patch does not have any anti-patterns.  |
|||| _ HBASE-29081 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 12s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   4m  3s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  compile  |   4m 45s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  checkstyle  |   1m  8s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  spotbugs  |   2m 39s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  spotless  |   0m 55s |  |  branch has no errors when running spotless:check.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 12s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   4m  2s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   4m 41s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   4m 41s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   1m 12s |  |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   2m 54s |  |  the patch passed  |
| +1 :green_heart: |  hadoopcheck  |  14m  7s |  |  Patch does not cause any errors with Hadoop 3.3.6 3.4.0.  |
| -1 :x: |  spotless  |   0m 19s |  |  patch has 21 errors when running spotless:check, run spotless:apply to fix.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 20s |  |  The patch does not generate ASF License warnings.  |
|  |   |  51m 20s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Optional Tests | dupname asflicense javac spotbugs checkstyle codespell detsecrets compile hadoopcheck hbaseanti spotless |
| uname | Linux 9c04949c24d7 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-29081 / 8f500f82c9a7d3dcb75b8917c8b731bc3125d983 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| spotless | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/1/artifact/yetus-general-check/output/patch-spotless.txt |
| Max. process+thread count | 84 (vs. ulimit of 30000) |
| modules | C: hbase-common hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/1/console |
| versions | git=2.34.1 maven=3.9.8 spotbugs=4.7.3 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 29s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ HBASE-29081 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 10s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 10s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  compile  |   1m 16s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  javadoc  |   0m 44s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  shadedjars  |   5m 58s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 12s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m  6s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 18s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 18s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 42s |  |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   5m 52s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  |   2m 57s | [/patch-unit-hbase-common.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/1/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-hbase-common.txt) |  hbase-common in the patch failed.  |
| -1 :x: |  unit  | 365m 46s | [/patch-unit-hbase-server.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/1/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-hbase-server.txt) |  hbase-server in the patch failed.  |
|  |   | 397m  5s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/1/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux af36b7a0be28 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-29081 / 8f500f82c9a7d3dcb75b8917c8b731bc3125d983 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/1/testReport/ |
| Max. process+thread count | 5150 (vs. ulimit of 30000) |
| modules | C: hbase-common hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/1/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m  0s |  |  Docker mode activated.  |
| -1 :x: |  patch  |   0m  7s |  |  https://github.com/apache/hbase/pull/6887 does not apply to HBASE-29081. Rebase required? Wrong Branch? See https://yetus.apache.org/documentation/in-progress/precommit-patchnames for help.  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/2/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m  0s |  |  Docker mode activated.  |
| -1 :x: |  patch  |   0m  7s |  |  https://github.com/apache/hbase/pull/6887 does not apply to HBASE-29081. Rebase required? Wrong Branch? See https://yetus.apache.org/documentation/in-progress/precommit-patchnames for help.  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/2/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

anmolnar commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/MasterFileSystem.java:
```diff
@@ -382,4 +392,106 @@ public void stop() {
+          acs.equals(getActiveClusterSuffix())
+            || (acs.toString().equals("null") && getActiveClusterSuffix().toString().isEmpty())
+        ) {
+          LOG.info("Active cluster is being restarted, Suffix {}", acs);
```
I'm not sure if 'restart' is always the case. Instead I'd log something like:
> I'm the active cluster on this storage location

anmolnar commented on hbase-protocol-shaded/src/main/protobuf/server/ActiveClusterSuffix.proto:
```diff
@@ -0,0 +1,33 @@
+ */
+message ActiveClusterSuffix {
+  // This is the active cluster suffix set by the user in the config, as a String
+  required string active_cluster_suffix = 1;
```
Are you going to store only the suffix?
Suffix is usually empty for the active cluster, so you'll write an empty file in that case. I'd suggest to add the cluster id as well: "clusterId_metasuffix".

kabhishek4 replied:
It can be stored as 2 fields but I have first coded considering 1 field after getting comments added the cluster id. 
Both the fields are string. For primary, the suffix is anyways null. Also, concatenated string comparison serves the purpose.

anmolnar commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/MasterFileSystem.java:
```diff
@@ -382,4 +392,106 @@ public void stop() {
+        // verify the contents against the config set
+        ActiveClusterSuffix acs = FSUtils.getActiveClusterSuffix(fs, rootdir);
+        if (
+          acs.equals(getActiveClusterSuffix())
+            || (acs.toString().equals("null") && getActiveClusterSuffix().toString().isEmpty())
```
We should add the clusterId to the file contents, in which case we could verify both: clusterId matches and suffix matches.

anmolnar commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/MasterFileSystem.java:
```diff
@@ -382,4 +392,106 @@ public void stop() {
+      }
+    } else {
+      // this is a replica cluster
+      if (!activeClusterSuffixFileExists) {
+        // suffix file must exist, if not throw error
+        LOG.info("rootdir {}", rootdir);
+        throw new IOException("Read Replica Cluster must have associated Active Cluster Suffix "
+          + "file. File 'active_cluster_suffix.id' does not exist in $HBASE_ROOT.");
+      }
```
This case doesn't necesserily have to be an error: users should be able to start the read replica clusters without and active cluster running. The new file is actually only important for non-RO mode clusters, RO instances don't need to check it at all.

anmolnar commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/MasterFileSystem.java:
```diff
@@ -382,4 +392,106 @@ public void stop() {
+
+  // This API should be to be called on the 'old active cluster' before the switch over,
+  // to delete the associated active_cluster.id file.
+  public boolean prepareForSwitchOver() {
```
Where do you use this method?

anmolnar commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/MasterFileSystem.java:
```diff
@@ -382,4 +392,106 @@ public void stop() {
+
+  // This API should be called on the 'new active cluster' to complete the process of switch over,
+  // to create the new active_cluster.id file.
+  public boolean performSwitchOver() {
```
Where do you use this method?

anmolnar commented on hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestClusterId.java:
```diff
@@ -114,5 +114,4 @@ public void testRewritingClusterIdToPB() throws Exception {
     HMaster master = TEST_UTIL.getHBaseCluster().getMaster();
     assertEquals(1, master.getServerManager().getOnlineServersList().size());
   }
-
```
Please remove this change.

anmolnar reviewed (changes requested):
Few comments.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 34s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  buf  |   0m  0s |  |  buf was not available.  |
| +0 :ok: |  buf  |   0m  0s |  |  buf was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  |  Patch does not have any anti-patterns.  |
|||| _ HBASE-29081 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 10s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 13s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  compile  |   5m  4s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  checkstyle  |   1m 11s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  spotbugs  |   4m 59s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  spotless  |   0m 45s |  |  branch has no errors when running spotless:check.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 11s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m  0s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   4m 57s |  |  the patch passed  |
| +1 :green_heart: |  cc  |   4m 57s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   4m 57s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   0m 35s | [/results-checkstyle-hbase-server.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/3/artifact/yetus-general-check/output/results-checkstyle-hbase-server.txt) |  hbase-server: The patch generated 1 new + 7 unchanged - 0 fixed = 8 total (was 7)  |
| -1 :x: |  spotbugs  |   1m 42s | [/new-spotbugs-hbase-server.html](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/3/artifact/yetus-general-check/output/new-spotbugs-hbase-server.html) |  hbase-server generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0)  |
| +1 :green_heart: |  hadoopcheck  |  11m 44s |  |  Patch does not cause any errors with Hadoop 3.3.6 3.4.0.  |
| +1 :green_heart: |  hbaseprotoc  |   1m 46s |  |  the patch passed  |
| +1 :green_heart: |  spotless  |   0m 43s |  |  patch has no errors when running spotless:check.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 34s |  |  The patch does not generate ASF License warnings.  |
|  |   |  53m 33s |  |  |


| Reason | Tests |
|-------:|:------|
| SpotBugs | module:hbase-server |
|  |  Call to String.equals(org.apache.hadoop.hbase.ActiveClusterSuffix) in org.apache.hadoop.hbase.master.MasterFileSystem.negotiateActiveClusterSuffixFile(long)  At MasterFileSystem.java: At MasterFileSystem.java:[line 411] |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/3/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Optional Tests | dupname asflicense javac spotbugs checkstyle codespell detsecrets compile hadoopcheck hbaseanti spotless cc buflint bufcompat hbaseprotoc |
| uname | Linux f5a6c2d7dc4a 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-29081 / 922815e199c01432829bb7cd1ee19fc82db9f34e |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| Max. process+thread count | 85 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-common hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/3/console |
| versions | git=2.34.1 maven=3.9.8 spotbugs=4.7.3 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 34s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ HBASE-29081 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 35s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 29s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  compile  |   2m 11s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  javadoc  |   1m  9s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  shadedjars  |   5m 55s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 13s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 59s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   2m  6s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   2m  6s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m  7s |  |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   5m 53s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 35s |  |  hbase-protocol-shaded in the patch passed.  |
| -1 :x: |  unit  |   3m  4s | [/patch-unit-hbase-common.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/3/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-hbase-common.txt) |  hbase-common in the patch failed.  |
| +1 :green_heart: |  unit  |   1m 34s |  |  hbase-client in the patch passed.  |
| -1 :x: |  unit  | 372m 45s | [/patch-unit-hbase-server.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/3/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-hbase-server.txt) |  hbase-server in the patch failed.  |
|  |   | 409m 41s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/3/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux f247271727b3 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-29081 / 922815e199c01432829bb7cd1ee19fc82db9f34e |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/3/testReport/ |
| Max. process+thread count | 4763 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-common hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/3/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

anmolnar commented on hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestActiveClusterSuffix.java:
```diff
@@ -0,0 +1,133 @@
+  public void testActiveClusterSuffixCreated() throws Exception {
+    TEST_UTIL.startMiniZKCluster();
+    TEST_UTIL.startMiniDFSCluster(1);
+    TEST_UTIL.createRootDir();
```
I don't think this is needed, since the next method (start mini cluster) will call it anyways.

anmolnar commented on hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestActiveClusterSuffix.java:
```diff
@@ -0,0 +1,133 @@
+  public void testSuffixFileOnRestart() throws Exception {
+    TEST_UTIL.startMiniZKCluster();
+    TEST_UTIL.startMiniDFSCluster(1);
+    TEST_UTIL.createRootDir();
```
Same here.

anmolnar commented on hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestActiveClusterSuffix.java:
```diff
@@ -0,0 +1,133 @@
+    FileSystem fs = rootDir.getFileSystem(TEST_UTIL.getConfiguration());
+    Path filePath = new Path(rootDir, HConstants.ACTIVE_CLUSTER_SUFFIX_FILE_NAME);
+
+    assertTrue(filePath + " should exists ", fs.exists(filePath));
+    assertTrue(filePath + " should not be empty  ", fs.getFileStatus(filePath).getLen() > 0);
```
This test should be the first positive case of the following two:
- start with empty root dir: active cluster id file created and contents verified by the test,
- start with existing root dir with existing active cluster id file, contents suitable for the cluster and starts up correctly.

anmolnar replied:
> I am getting error if I stop and start mini cluster in the same test case hence distributed the scenarios.

That's very good idea.

anmolnar commented on hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestActiveClusterSuffix.java:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  @Test
+  public void testSuffixFileOnRestart() throws Exception {
```
This test could be the 2nd of two positive test case mentioned previously.

anmolnar commented on hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestActiveClusterSuffix.java:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  @Test
+  public void testVerifyErrorWhenSuffixNotMatched() throws Exception {
```
This one is the only negative case and fine as it is:
- start with existing root with active cluster id file for a different cluster.

anmolnar reviewed (changes requested):
Please review my suggestions for the test cases.

kabhishek4 commented:
Changed code as per comments. Also, found a bug during testing fixed it where activeClusterSuffix is now initialised while writing suffix in the file and reading from it. Removed unwanted call to rewriteAsPb as well.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 31s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ HBASE-29081 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 10s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 17s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  compile  |   2m  8s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  javadoc  |   1m 10s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  shadedjars  |   5m 57s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 12s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m  6s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   2m  8s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   2m  8s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m  8s |  |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   5m 54s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 35s |  |  hbase-protocol-shaded in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 17s |  |  hbase-common in the patch passed.  |
| +1 :green_heart: |  unit  |   1m 35s |  |  hbase-client in the patch passed.  |
| -1 :x: |  unit  | 225m 45s | [/patch-unit-hbase-server.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/4/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-hbase-server.txt) |  hbase-server in the patch failed.  |
|  |   | 262m  1s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/4/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux 72fd5cb3f5d0 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-29081 / 01430d1efcfc2a2cfccc6a2091cae60f9ff0831c |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/4/testReport/ |
| Max. process+thread count | 5259 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-common hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/4/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 29s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  buf  |   0m  0s |  |  buf was not available.  |
| +0 :ok: |  buf  |   0m  0s |  |  buf was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  |  Patch does not have any anti-patterns.  |
|||| _ HBASE-29081 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 40s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   4m 31s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  compile  |   6m 49s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  checkstyle  |   1m 45s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  spotbugs  |   7m  1s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  spotless  |   1m  4s |  |  branch has no errors when running spotless:check.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 12s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   4m 11s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   7m 31s |  |  the patch passed  |
| +1 :green_heart: |  cc  |   7m 31s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   7m 31s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   2m 15s |  |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   7m 20s |  |  the patch passed  |
| +1 :green_heart: |  hadoopcheck  |  13m  1s |  |  Patch does not cause any errors with Hadoop 3.3.6 3.4.0.  |
| +1 :green_heart: |  hbaseprotoc  |   2m  1s |  |  the patch passed  |
| +1 :green_heart: |  spotless  |   0m 50s |  |  patch has no errors when running spotless:check.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 37s |  |  The patch does not generate ASF License warnings.  |
|  |   |  69m 35s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/5/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Optional Tests | dupname asflicense javac spotbugs checkstyle codespell detsecrets compile hadoopcheck hbaseanti spotless cc buflint bufcompat hbaseprotoc |
| uname | Linux f718bb9270cc 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-29081 / 39f8d25820d812da1ae91143f9ba7ecbec78ff9e |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| Max. process+thread count | 84 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-common hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/5/console |
| versions | git=2.34.1 maven=3.9.8 spotbugs=4.7.3 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 31s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ HBASE-29081 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 38s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 51s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  compile  |   3m 16s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  javadoc  |   1m 44s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  shadedjars  |   7m 48s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   4m 13s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   2m 58s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   2m 58s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m 40s |  |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   7m 29s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 40s |  |  hbase-protocol-shaded in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 51s |  |  hbase-common in the patch passed.  |
| +1 :green_heart: |  unit  |   2m  0s |  |  hbase-client in the patch passed.  |
| -1 :x: |  unit  | 276m 36s | [/patch-unit-hbase-server.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/5/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-hbase-server.txt) |  hbase-server in the patch failed.  |
|  |   | 322m 52s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/5/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux ac7eed4ab686 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-29081 / 39f8d25820d812da1ae91143f9ba7ecbec78ff9e |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/5/testReport/ |
| Max. process+thread count | 5217 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-common hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/5/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

anmolnar commented on hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestActiveClusterSuffix.java:
```diff
@@ -0,0 +1,161 @@
+    } catch (IOException ioe) {
+      threwIOE = true;
+    } finally {
+      assertFalse("The master should not have thrown an exception", threwIOE);
```
Don't need a variable for this. You can do `fail(...)` in the catch clause.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 33s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  buf  |   0m  0s |  |  buf was not available.  |
| +0 :ok: |  buf  |   0m  0s |  |  buf was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  |  Patch does not have any anti-patterns.  |
|||| _ HBASE-29081 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 12s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 16s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  compile  |   5m 21s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  checkstyle  |   1m 12s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  spotbugs  |   5m  9s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  spotless  |   0m 47s |  |  branch has no errors when running spotless:check.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 11s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m  4s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   7m 16s |  |  the patch passed  |
| +1 :green_heart: |  cc  |   7m 16s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   7m 16s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   1m 36s |  |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   5m 53s |  |  the patch passed  |
| +1 :green_heart: |  hadoopcheck  |  16m 53s |  |  Patch does not cause any errors with Hadoop 3.3.6 3.4.0.  |
| +1 :green_heart: |  hbaseprotoc  |   9m 44s |  |  the patch passed  |
| -1 :x: |  spotless  |   1m 25s |  |  patch has 1 errors when running spotless:check, run spotless:apply to fix.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 49s |  |  The patch does not generate ASF License warnings.  |
|  |   |  76m  8s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/6/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Optional Tests | dupname asflicense javac spotbugs checkstyle codespell detsecrets compile hadoopcheck hbaseanti spotless cc buflint bufcompat hbaseprotoc |
| uname | Linux 5b60b26832c8 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-29081 / 39f8d25820d812da1ae91143f9ba7ecbec78ff9e |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| spotless | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/6/artifact/yetus-general-check/output/patch-spotless.txt |
| Max. process+thread count | 85 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-common hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/6/console |
| versions | git=2.34.1 maven=3.9.8 spotbugs=4.7.3 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 30s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ HBASE-29081 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 10s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 10s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  compile  |   2m 10s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  javadoc  |   1m  9s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  shadedjars  |   5m 57s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 13s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m  2s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   2m 10s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   2m 10s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m  9s |  |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   5m 55s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 33s |  |  hbase-protocol-shaded in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 14s |  |  hbase-common in the patch passed.  |
| +1 :green_heart: |  unit  |   1m 35s |  |  hbase-client in the patch passed.  |
| -1 :x: |  unit  | 229m 17s | [/patch-unit-hbase-server.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/6/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-hbase-server.txt) |  hbase-server in the patch failed.  |
|  |   | 264m 37s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/6/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux e1f856d19629 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-29081 / 39f8d25820d812da1ae91143f9ba7ecbec78ff9e |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/6/testReport/ |
| Max. process+thread count | 5430 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-common hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/6/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 47s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  buf  |   0m  0s |  |  buf was not available.  |
| +0 :ok: |  buf  |   0m  0s |  |  buf was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  |  Patch does not have any anti-patterns.  |
|||| _ HBASE-29081 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 38s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   5m 25s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  compile  |   7m 12s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  checkstyle  |   1m 36s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  spotbugs  |   7m 36s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  spotless  |   1m 17s |  |  branch has no errors when running spotless:check.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 18s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   5m  2s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   7m 37s |  |  the patch passed  |
| +1 :green_heart: |  cc  |   7m 37s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   7m 37s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   1m 56s |  |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   8m 28s |  |  the patch passed  |
| +1 :green_heart: |  hadoopcheck  |  14m 32s |  |  Patch does not cause any errors with Hadoop 3.3.6 3.4.0.  |
| +1 :green_heart: |  hbaseprotoc  |   2m 42s |  |  the patch passed  |
| +1 :green_heart: |  spotless  |   0m 55s |  |  patch has no errors when running spotless:check.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 53s |  |  The patch does not generate ASF License warnings.  |
|  |   |  76m 42s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/7/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Optional Tests | dupname asflicense javac spotbugs checkstyle codespell detsecrets compile hadoopcheck hbaseanti spotless cc buflint bufcompat hbaseprotoc |
| uname | Linux 837e2ba3b1b0 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-29081 / a2e4b9cb7c5e9dbc7d58525ca95e8c91c639a018 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| Max. process+thread count | 84 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-common hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/7/console |
| versions | git=2.34.1 maven=3.9.8 spotbugs=4.7.3 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |  19m 10s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m 14s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ HBASE-29081 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 45s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  10m  8s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  compile  |   3m 27s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  javadoc  |   1m 46s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  shadedjars  |   8m 48s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 15s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   6m 41s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   5m 24s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   5m 24s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m 33s |  |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   7m 15s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 47s |  |  hbase-protocol-shaded in the patch passed.  |
| +1 :green_heart: |  unit  |   3m 18s |  |  hbase-common in the patch passed.  |
| +1 :green_heart: |  unit  |   2m  0s |  |  hbase-client in the patch passed.  |
| -1 :x: |  unit  | 301m 58s | [/patch-unit-hbase-server.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/7/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-hbase-server.txt) |  hbase-server in the patch failed.  |
|  |   | 382m 53s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/7/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux c685939563a9 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-29081 / a2e4b9cb7c5e9dbc7d58525ca95e8c91c639a018 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/7/testReport/ |
| Max. process+thread count | 5174 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-common hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/7/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

anmolnar commented:
@wchevreuil You might want to check these test failure, because as far as I can see flaky test fixes are already rebased to the feature branch.

kabhishek4 commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/MasterFileSystem.java:
```diff
@@ -382,4 +393,68 @@ public void stop() {
+
+  private void negotiateActiveClusterSuffixFile(long wait) throws IOException {
+    String suffix_file_name = HConstants.ACTIVE_CLUSTER_SUFFIX_FILE_NAME;
+    boolean activeClusterSuffixFileExists =
+      FSUtils.checkFileExistsInHbaseRootDir(fs, rootdir, suffix_file_name, wait);
```
Yes. I think it is possible to use exception types coming from getActiveClusterSuffix method and update the logic here to do it one step. I will try out the changes and update the patch.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   3m 24s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ HBASE-29081 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 11s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 15s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  compile  |   2m 15s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  javadoc  |   1m 12s |  |  HBASE-29081 passed  |
| +1 :green_heart: |  shadedjars  |   6m  2s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 13s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m  8s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   2m 12s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   2m 12s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m  9s |  |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   5m 57s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 33s |  |  hbase-protocol-shaded in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 17s |  |  hbase-common in the patch passed.  |
| +1 :green_heart: |  unit  |   1m 36s |  |  hbase-client in the patch passed.  |
| -1 :x: |  unit  | 256m 49s | [/patch-unit-hbase-server.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/9/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-hbase-server.txt) |  hbase-server in the patch failed.  |
|  |   | 296m 20s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/9/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6887 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux 9562ddbc6718 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-29081 / 481e4e6045f43721e5e69537ab3d1877c7c17006 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/9/testReport/ |
| Max. process+thread count | 4528 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-common hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6887/9/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

anmolnar commented:
> active cluster startup attempted after setting hbase.global.readonly.enabled to true

What do you mean by "active cluster startup attempted"?
If readonly mode is true, it means user attempts to start a replica cluster which will succeed always, because in this case we don't check the active cluster id file.

Similarly

> replica cluster startup attempted after setting hbase.global.readonly.enabled to false

This is going to be an active cluster startup attempt which will succeed only if:
- active cluster id file is missing
- active cluster id file contents matches our suffix.

kgeisz commented:
> What do you mean by "active cluster startup attempted"?

I got it from the design document:

> When the ‘active_cluster_id’ file is created on cloud storage with cluster id of active cluster, startup will not be allowed and error will be reported in below scenarios,
> - active cluster startup attempted after setting hbase.global.readonly.enabled to true
> - replica cluster startup attempted after setting hbase.global.readonly.enabled to false

My understanding is "active cluster startup attempted" means the cluster whose ID and suffix is in the `active_cluster_id` file is the cluster that's being started.

> If readonly mode is true, it means user attempts to start a replica cluster which will succeed always, because in this case we don't check the active cluster id file.

That makes sense, although it doesn't align with what the design doc says.  I was expecting this scenario to result in an exception.
